recipes/moltenvk/all/conanfile.py

```diff
@@ -84,63 +97,74 @@ def _require(self, recipe_name):
     def package_id(self):
         # MoltenVK >=1.O.42 requires at least XCode 12.0 (11.4 actually) at build
         # time but can be consumed by older compiler versions if shared
-        if tools.Version(self.version) >= "1.0.42" and self.options.shared:
-            if tools.Version(self.settings.compiler.version) < "12.0":
+        if Version(self.version) >= "1.0.42" and self.options.shared:
+            if Version(self.settings.compiler.version) < "12.0":
                 compatible_pkg = self.info.clone()
                 compatible_pkg.settings.compiler.version = "12.0"
                 self.compatible_packages.append(compatible_pkg)
 
     def validate(self):
-        if self.settings.compiler.get_safe("cppstd"):
-            tools.check_min_cppstd(self, self._min_cppstd)
-        if self.settings.os not in ["Macos", "iOS", "tvOS"]:
-            raise ConanInvalidConfiguration("MoltenVK only supported on MacOS, iOS and tvOS")
-        if self.settings.compiler != "apple-clang":
-            raise ConanInvalidConfiguration("MoltenVK requires apple-clang")
-        if tools.Version(self.version) >= "1.0.42":
-            if tools.Version(self.settings.compiler.version) < "12.0":
-                raise ConanInvalidConfiguration("MoltenVK {} requires XCode 12.0 or higher at build time".format(self.version))
+        if self.info.settings.compiler.get_safe("cppstd"):
+            check_min_cppstd(self, self._min_cppstd)
+        if self.info.settings.os not in ["Macos", "iOS", "tvOS"]:
+            raise ConanInvalidConfiguration(f"{self.ref} only supported on MacOS, iOS and tvOS")
+        if self.info.settings.compiler != "apple-clang":
+            raise ConanInvalidConfiguration(f"{self.ref} requires apple-clang")
+        if Version(self.version) >= "1.0.42" and not self.info.options.shared:
+            if Version(self.info.settings.compiler.version) < "12.0":
+                raise ConanInvalidConfiguration(f"{self.ref} static requires XCode 12.0 or higher at build & consume time")
+
+    def validate_build(self):
+        if Version(self.version) >= "1.0.42":
+            if Version(self.settings.compiler.version) < "12.0":
+                raise ConanInvalidConfiguration(f"{self.ref} requires XCode 12.0 or higher at build time")
+        spirv_cross = self.dependencies["spirv-cross"]
+        if spirv_cross.options.shared or not spirv_cross.options.msl or not spirv_cross.options.reflect:
+            raise ConanInvalidConfiguration(f"{self.ref} requires spirv-cross static with msl & reflect enabled")
```

Comment:
@jgsogo c3i doesn't seem to be smart enough to handle this logic.

```
ERROR: Missing binary: moltenvk/1.1.11:b1ad2d34d7dbcd0375bec2fba400e59367e8d4da
moltenvk/1.1.11: WARN: Can't find a 'moltenvk/1.1.11' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, build_type=Debug, compiler=apple-clang, compiler.libcxx=libc++, compiler.version=11.0, os=Macos
- Options: hide_vulkan_symbols=False, shared=True, tools=True, with_spirv_tools=True, cereal:thread_safe=False, glslang:build_executables=True, glslang:enable_optimizer=True, glslang:fPIC=True, glslang:hlsl=True, glslang:shared=False, glslang:spv_remapper=True, spirv-cross:build_executable=True, spirv-cross:c_api=True, spirv-cross:cpp=True, spirv-cross:exceptions=True, spirv-cross:fPIC=True, spirv-cross:glsl=True, spirv-cross:hlsl=True, spirv-cross:msl=True, spirv-cross:namespace=spirv_cross, spirv-cross:reflect=True, spirv-cross:shared=False, spirv-cross:util=True, spirv-tools:build_executables=True, spirv-tools:fPIC=True, spirv-tools:shared=False
- Dependencies: cereal/1.3.2, glslang/1.3.224.0, spirv-cross/1.3.224.0, vulkan-headers/1.3.224.0, spirv-tools/1.3.224.0
- Requirements: cereal/1.Y.Z, glslang/1.Y.Z, spirv-cross/1.Y.Z, spirv-tools/1.Y.Z, vulkan-headers/1.Y.Z
- Package ID: b1ad2d34d7dbcd0375bec2fba400e59367e8d4da

ERROR: Missing prebuilt package for 'moltenvk/1.1.11'
Use 'conan search moltenvk/1.1.11 --table=table.html -r=remote' and open the table.html file to see available packages
Or try to build locally from sources with '--build=moltenvk'
```

Reply:
Basically it should not try to download an existing package for clang 11 if static, but build from source.

Reply:
The error doesn't make any sense here. The purpose of the command is to build that same package `conan install moltenvk/1.1.11@#aaf875be6b82652365ed2e0bdc85f091 --build=moltenvk`, so it cannot complain about it being missing 😵‍💫  Is this a Conan bug?

Reply:
Maybe some weird effect due to compatible pkg, validate & validate_build:

```python
    def package_id(self):
        # MoltenVK >=1.O.42 requires at least XCode 12.0 (11.4 actually) at build
        # time but can be consumed by older compiler versions if shared
        if Version(self.version) >= "1.0.42" and self.options.shared:
            if Version(self.settings.compiler.version) < "12.0":
                compatible_pkg = self.info.clone()
                compatible_pkg.settings.compiler.version = "12.0"
                self.compatible_packages.append(compatible_pkg)

    def validate(self):
        if self.info.settings.compiler.get_safe("cppstd"):
            check_min_cppstd(self, self._min_cppstd)
        if self.info.settings.os not in ["Macos", "iOS", "tvOS"]:
            raise ConanInvalidConfiguration(f"{self.ref} only supported on MacOS, iOS and tvOS")
        if self.info.settings.compiler != "apple-clang":
            raise ConanInvalidConfiguration(f"{self.ref} requires apple-clang")
        if Version(self.version) >= "1.0.42" and not self.info.options.shared:
            if Version(self.info.settings.compiler.version) < "12.0":
                raise ConanInvalidConfiguration(f"{self.ref} static requires XCode 12.0 or higher at build & consume time")

    def validate_build(self):
        if Version(self.version) >= "1.0.42":
            if Version(self.settings.compiler.version) < "12.0":
                raise ConanInvalidConfiguration(f"{self.ref} requires XCode 12.0 or higher at build time")
        spirv_cross = self.dependencies["spirv-cross"]
        if spirv_cross.options.shared or not spirv_cross.options.msl or not spirv_cross.options.reflect:
            raise ConanInvalidConfiguration(f"{self.ref} requires spirv-cross static with msl & reflect enabled")
```

Reply:
So what can I do here?